File: src/test/helper/testStorage/TestStorage.cpp
#include "TestStorage.h"

#include "AccessKind.h"
#include "Edge.h"
#include "FilePath.h"
#include "LocationType.h"
#include "NameHierarchy.h"
#include "NodeType.h"
#include "Storage.h"
#include "utilityString.h"

std::shared_ptr<TestStorage> TestStorage::create(const std::shared_ptr<const Storage>& storage) {
  auto testStorage = std::make_shared<TestStorage>();

  std::map<Id, FilePath> filePathMap;
  for(const auto& file : storage->getStorageFiles()) {
    filePathMap.emplace(file.id, FilePath(file.filePath));
    testStorage->files.emplace(file.filePath);

    testStorage->addLine(L"FILE: " + FilePath(file.filePath).fileName() + (file.indexed ? L"" : L" non-indexed"));
  }

  std::map<Id, StorageComponentAccess> accessMap;
  for(const auto& access : storage->getComponentAccesses()) {
    accessMap.emplace(access.nodeId, access);
  }

  std::multimap<Id, Id> occurrenceMap;
  for(const auto& occurrence : storage->getStorageOccurrences()) {
    occurrenceMap.emplace(occurrence.sourceLocationId, occurrence.elementId);
  }

  std::multimap<Id, StorageSourceLocation> tokenLocationMap;
  std::multimap<Id, StorageSourceLocation> scopeLocationMap;
  std::multimap<Id, StorageSourceLocation> signatureLocationMap;
  std::multimap<Id, StorageSourceLocation> localSymbolLocationMap;
  std::multimap<Id, StorageSourceLocation> qualifierLocationMap;
  std::multimap<Id, StorageSourceLocation> errorLocationMap;
  std::vector<StorageSourceLocation> commentLocations;
  for(const auto& location : storage->getStorageSourceLocations()) {
    std::vector<Id> elementIds;
    for(auto it = occurrenceMap.find(location.id); it != occurrenceMap.end() && it->first == location.id; it++) {
      elementIds.emplace_back(it->second);
    }

    if(elementIds.empty()) {
      elementIds.emplace_back(0);
    }

    for(Id elementId : elementIds) {
      switch(intToLocationType(location.type)) {
      case LOCATION_TOKEN:
        if(elementId != 0U) {
          tokenLocationMap.emplace(elementId, location);
        }
        break;
      case LOCATION_SCOPE:
        if(elementId != 0U) {
          scopeLocationMap.emplace(elementId, location);
        }
        break;
      case LOCATION_QUALIFIER:
        if(elementId != 0U) {
          qualifierLocationMap.emplace(elementId, location);
        }
        break;
      case LOCATION_LOCAL_SYMBOL:
        if(elementId != 0U) {
          localSymbolLocationMap.emplace(elementId, location);
        }
        break;
      case LOCATION_SIGNATURE:
        if(elementId != 0U) {
          signatureLocationMap.emplace(elementId, location);
        }
        break;
      case LOCATION_ERROR:
        if(elementId != 0U) {
          errorLocationMap.emplace(elementId, location);
        }
        break;
      case LOCATION_COMMENT:
        commentLocations.emplace_back(location);
        break;
      default:
        break;
      }
    }
  }

  std::map<Id, StorageNode> nodesMap;
  std::set<Id> fileIdMap;
  for(const auto& node : storage->getStorageNodes()) {
    nodesMap.emplace(node.id, node);

    if(intToNodeKind(node.type) == NODE_FILE) {
      fileIdMap.insert(node.id);
    }

    std::wstring nameStr = NameHierarchy::deserialize(node.serializedName).getQualifiedNameWithSignature();

    for(auto qualifierLocationIt = qualifierLocationMap.find(node.id);
        qualifierLocationIt != qualifierLocationMap.end() && qualifierLocationIt->first == node.id;
        qualifierLocationIt++) {
      std::wstring locStr = addLocationStr(L"", qualifierLocationIt->second);
      testStorage->qualifiers.emplace_back(nameStr + locStr);
      testStorage->addLine(L"QUALIFIER: " + nameStr + addFileName(locStr, filePathMap[qualifierLocationIt->second.fileNodeId]));
    }

    auto* bin = testStorage->getBinForNodeType(node.type);
    if(bin != nullptr) {
      auto accessIt = accessMap.find(node.id);
      if(accessIt != accessMap.end()) {
        if(intToAccessKind(accessIt->second.type) != ACCESS_TEMPLATE_PARAMETER &&
           intToAccessKind(accessIt->second.type) != ACCESS_TYPE_PARAMETER) {
          nameStr = accessKindToString(intToAccessKind(accessIt->second.type)) + L' ' + nameStr;
        }
      }

      bool added = false;
      for(auto tokenLocationIt = tokenLocationMap.find(node.id);
          tokenLocationIt != tokenLocationMap.end() && tokenLocationIt->first == node.id;
          tokenLocationIt++) {
        added = false;
        std::wstring locationStr = addLocationStr(L"", tokenLocationIt->second);

        auto signatureLocationIt = signatureLocationMap.find(node.id);
        if(signatureLocationIt != signatureLocationMap.end() &&
           containsLocation(signatureLocationIt->second, tokenLocationIt->second)) {
          locationStr = addLocationStr(locationStr, signatureLocationIt->second);
        }

        for(auto scopeLocationIt = scopeLocationMap.find(node.id);
            scopeLocationIt != scopeLocationMap.end() && scopeLocationIt->first == node.id;
            scopeLocationIt++) {
          if(containsLocation(scopeLocationIt->second, tokenLocationIt->second)) {
            std::wstring locStr = addLocationStr(locationStr, scopeLocationIt->second);
            bin->emplace_back(nameStr + locStr);
            testStorage->addLine(nodeTypeToString(node.type) + L": " + nameStr +
                                 addFileName(locStr, filePathMap[tokenLocationIt->second.fileNodeId]));
            added = true;
          }
        }

        if(!added) {
          bin->emplace_back(nameStr + locationStr);
          testStorage->addLine(nodeTypeToString(node.type) + L": " + nameStr +
                               addFileName(locationStr, filePathMap[tokenLocationIt->second.fileNodeId]));
          added = true;
        }
      }

      if(!added) {
        bin->emplace_back(nameStr);
        testStorage->addLine(nodeTypeToString(node.type) + L": " + nameStr);
      }
    }
  }

  for(const auto& edge : storage->getStorageEdges()) {
    auto targetIt = nodesMap.find(edge.targetNodeId);
    if(targetIt == nodesMap.end()) {
      continue;
    }

    auto sourceIt = nodesMap.find(edge.sourceNodeId);
    if(sourceIt == nodesMap.end()) {
      continue;
    }

    const StorageNode& target = targetIt->second;
    const StorageNode& source = sourceIt->second;

    std::vector<std::wstring>* bin = testStorage->getBinForEdgeType(edge.type);
    if(bin != nullptr) {
      std::wstring sourceName = NameHierarchy::deserialize(source.serializedName).getQualifiedNameWithSignature();
      if(fileIdMap.find(edge.sourceNodeId) != fileIdMap.end() && FilePath(sourceName).exists()) {
        sourceName = FilePath(sourceName).fileName();
      }

      std::wstring targetName = NameHierarchy::deserialize(target.serializedName).getQualifiedNameWithSignature();
      if(fileIdMap.find(edge.targetNodeId) != fileIdMap.end() && FilePath(targetName).exists()) {
        targetName = FilePath(targetName).fileName();
      }

      std::wstring nameStr = sourceName + L" -> " + targetName;

      bool added = false;
      for(auto tokenLocationIt = tokenLocationMap.find(edge.id);
          tokenLocationIt != tokenLocationMap.end() && tokenLocationIt->first == edge.id;
          tokenLocationIt++) {
        std::wstring locStr = addLocationStr(L"", tokenLocationIt->second);
        bin->emplace_back(nameStr + locStr);
        testStorage->addLine(edgeTypeToString(edge.type) + L": " + nameStr +
                             addFileName(locStr, filePathMap[tokenLocationIt->second.fileNodeId]));
        added = true;
      }

      if(!added) {
        bin->emplace_back(nameStr);
        testStorage->addLine(edgeTypeToString(edge.type) + L": " + nameStr);
      }
    }
  }

  for(const auto& localSymbol : storage->getStorageLocalSymbols()) {
    bool added = false;
    for(auto localSymbolLocationIt = localSymbolLocationMap.find(localSymbol.id);
        localSymbolLocationIt != localSymbolLocationMap.end() && localSymbolLocationIt->first == localSymbol.id;
        localSymbolLocationIt++) {
      std::wstring locStr = addLocationStr(L"", localSymbolLocationIt->second);
      testStorage->localSymbols.emplace_back(localSymbol.name + locStr);
      testStorage->addLine(L"LOCAL_SYMBOL: " + localSymbol.name +
                           addFileName(locStr, filePathMap[localSymbolLocationIt->second.fileNodeId]));
      added = true;
    }

    if(!added) {
      testStorage->localSymbols.emplace_back(localSymbol.name);
      testStorage->addLine(L"LOCAL_SYMBOL: " + localSymbol.name);
    }
  }

  for(const auto& location : commentLocations) {
    std::wstring locStr = addLocationStr(L"", location);
    testStorage->comments.emplace_back(L"comment" + locStr);
    testStorage->addLine(L"COMMENT: comment" + addFileName(locStr, filePathMap[location.fileNodeId]));
  }

  for(const auto& error : storage->getErrors()) {
    for(auto errorLocationIt = errorLocationMap.find(error.id);
        errorLocationIt != errorLocationMap.end() && errorLocationIt->first == error.id;
        errorLocationIt++) {
      const auto& locStr = addLocationStr(L"", errorLocationIt->second);
      testStorage->errors.emplace_back(error.message + locStr);
      testStorage->addLine(L"ERROR: " + error.message + addFileName(locStr, filePathMap[errorLocationIt->second.fileNodeId]));
    }
  }

  return testStorage;
}

std::wstring TestStorage::nodeTypeToString(int nodeType) {
  switch(intToNodeKind(nodeType)) {
  case NODE_BUILTIN_TYPE:
    return L"SYMBOL_BUILTIN_TYPE";
  case NODE_CLASS:
    return L"SYMBOL_CLASS";
  case NODE_ENUM:
    return L"SYMBOL_ENUM";
  case NODE_ENUM_CONSTANT:
    return L"SYMBOL_ENUM_CONSTANT";
  case NODE_FIELD:
    return L"SYMBOL_FIELD";
  case NODE_FUNCTION:
    return L"SYMBOL_FUNCTION";
  case NODE_GLOBAL_VARIABLE:
    return L"SYMBOL_GLOBAL_VARIABLE";
  case NODE_INTERFACE:
    return L"SYMBOL_INTERFACE";
  case NODE_MACRO:
    return L"SYMBOL_MACRO";
  case NODE_METHOD:
    return L"SYMBOL_METHOD";
  case NODE_MODULE:
    return L"SYMBOL_MODULE";
  case NODE_NAMESPACE:
    return L"SYMBOL_NAMESPACE";
  case NODE_PACKAGE:
    return L"SYMBOL_PACKAGE";
  case NODE_STRUCT:
    return L"SYMBOL_STRUCT";
  case NODE_TYPEDEF:
    return L"SYMBOL_TYPEDEF";
  case NODE_TYPE_PARAMETER:
    return L"SYMBOL_TYPE_PARAMETER";
  case NODE_UNION:
    return L"SYMBOL_UNION";
  default:
    break;
  }
  return L"SYMBOL_NON_INDEXED";
}

std::wstring TestStorage::edgeTypeToString(int edgeType) {
  switch(Edge::intToType(edgeType)) {
  case Edge::EDGE_TYPE_USAGE:
    return L"REFERENCE_TYPE_USAGE";
  case Edge::EDGE_USAGE:
    return L"REFERENCE_USAGE";
  case Edge::EDGE_CALL:
    return L"REFERENCE_CALL";
  case Edge::EDGE_INHERITANCE:
    return L"REFERENCE_INHERITANCE";
  case Edge::EDGE_OVERRIDE:
    return L"REFERENCE_OVERRIDE";
  case Edge::EDGE_TYPE_ARGUMENT:
    return L"REFERENCE_TYPE_ARGUMENT";
  case Edge::EDGE_TEMPLATE_SPECIALIZATION:
    return L"REFERENCE_TEMPLATE_SPECIALIZATION";
  case Edge::EDGE_INCLUDE:
    return L"REFERENCE_INCLUDE";
  case Edge::EDGE_IMPORT:
    return L"REFERENCE_IMPORT";
  case Edge::EDGE_MACRO_USAGE:
    return L"REFERENCE_MACRO_USAGE";
  case Edge::EDGE_ANNOTATION_USAGE:
    return L"REFERENCE_ANNOTATION_USAGE";
  default:
    break;
  }
  return L"REFERENCE_UNDEFINED";
}

std::vector<std::wstring>* TestStorage::getBinForNodeType(int nodeType) {
  switch(intToNodeKind(nodeType)) {
  case NODE_PACKAGE:
    return &packages;
  case NODE_TYPEDEF:
    return &typedefs;
  case NODE_BUILTIN_TYPE:
    return &builtinTypes;
  case NODE_CLASS:
    return &classes;
  case NODE_UNION:
    return &unions;
  case NODE_INTERFACE:
    return &interfaces;
  case NODE_ANNOTATION:
    return &annotations;
  case NODE_ENUM:
    return &enums;
  case NODE_ENUM_CONSTANT:
    return &enumConstants;
  case NODE_FUNCTION:
    return &functions;
  case NODE_FIELD:
    return &fields;
  case NODE_GLOBAL_VARIABLE:
    return &globalVariables;
  case NODE_METHOD:
    return &methods;
  case NODE_MODULE:
    return &modules;
  case NODE_NAMESPACE:
    return &namespaces;
  case NODE_STRUCT:
    return &structs;
  case NODE_MACRO:
    return &macros;
  case NODE_TYPE_PARAMETER:
    return &typeParameters;
  default:
    break;
  }
  return nullptr;
}

std::vector<std::wstring>* TestStorage::getBinForEdgeType(int edgeType) {
  switch(Edge::intToType(edgeType)) {
  case Edge::EDGE_TYPE_USAGE:
    return &typeUses;
  case Edge::EDGE_USAGE:
    return &usages;
  case Edge::EDGE_CALL:
    return &calls;
  case Edge::EDGE_INHERITANCE:
    return &inheritances;
  case Edge::EDGE_OVERRIDE:
    return &overrides;
  case Edge::EDGE_TYPE_ARGUMENT:
    return &typeArguments;
  case Edge::EDGE_TEMPLATE_SPECIALIZATION:
    return &templateSpecializations;
  case Edge::EDGE_INCLUDE:
    return &includes;
  case Edge::EDGE_IMPORT:
    return &imports;
  case Edge::EDGE_MACRO_USAGE:
    return &macroUses;
  case Edge::EDGE_ANNOTATION_USAGE:
    return &annotationUses;
  default:
    break;
  }
  return nullptr;
}

std::wstring TestStorage::addLocationStr(const std::wstring& locationStr, const StorageSourceLocation& loc) {
  return L" <" + std::to_wstring(loc.startLine) + L':' + std::to_wstring(loc.startCol) + locationStr + L' ' +
      std::to_wstring(loc.endLine) + L':' + std::to_wstring(loc.endCol) + L'>';
}

std::wstring TestStorage::addFileName(const std::wstring& locationStr, const FilePath& filePath) {
  return L" [" + filePath.fileName() + locationStr + L']';
}

bool TestStorage::containsLocation(const StorageSourceLocation& outLocation, const StorageSourceLocation& inLocation) {
  if(outLocation.startLine > inLocation.startLine) {
    return false;
  }

  if(outLocation.startLine == inLocation.startLine && outLocation.startCol > inLocation.startCol) {
    return false;
  }

  if(outLocation.endLine < inLocation.endLine) {
    return false;
  }

  if(outLocation.endLine == inLocation.endLine && outLocation.endCol < inLocation.endCol) {
    return false;
  }

  return true;
}

void TestStorage::addLine(const std::wstring& message) {
  m_lines.emplace_back(utility::encodeToUtf8(message) + '\n');
}

File: indexers/java/indexer/src/test/java/com/sourcetrail/indexer/TestStorage.java
package com.sourcetrail.indexer;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import sourcetrail.SourcetrailCommon.IntermediateStorage;
import sourcetrail.SourcetrailCommon.StorageComponentAccess;
import sourcetrail.SourcetrailCommon.StorageEdge;
import sourcetrail.SourcetrailCommon.StorageError;
import sourcetrail.SourcetrailCommon.StorageFile;
import sourcetrail.SourcetrailCommon.StorageLocalSymbol;
import sourcetrail.SourcetrailCommon.StorageNode;
import sourcetrail.SourcetrailCommon.StorageOccurrence;
import sourcetrail.SourcetrailCommon.StorageSourceLocation;

/**
 * Java port of the C++ test helper {@code TestStorage}
 * ({@code src/lib/lib/tests/helper/testStorage/TestStorage.cpp}/{@code .h}).
 *
 * <p>The engine's own integration tests flatten a {@code Storage} into short, human-readable
 * strings such as {@code "public void Foo.m() <3:8 3:9>"} instead of asserting on raw ids. This
 * class produces the same strings from an {@code IntermediateStorage} proto so Java indexer tests
 * can read the same way (see {@code tests/integration/lib_cxx/CxxParser11TestSuite.cpp} for the
 * C++-side style this mirrors). Every formatting rule here (name rendering, access prefixing,
 * location nesting, node/edge bin routing) ports the C++ source line for line; read that file
 * before changing this one.
 */
public final class TestStorage {
  private final IntermediateStorage proto;

  public final List<String> packages = new ArrayList<>();
  public final List<String> classes = new ArrayList<>();
  public final List<String> interfaces = new ArrayList<>();
  public final List<String> annotations = new ArrayList<>();
  public final List<String> enums = new ArrayList<>();
  public final List<String> enumConstants = new ArrayList<>();
  public final List<String> methods = new ArrayList<>();
  public final List<String> fields = new ArrayList<>();
  public final List<String> functions = new ArrayList<>();
  public final List<String> typeParameters = new ArrayList<>();
  public final List<String> namespaces = new ArrayList<>();
  public final List<String> localSymbols = new ArrayList<>();
  public final List<String> qualifiers = new ArrayList<>();

  public final List<String> calls = new ArrayList<>();
  public final List<String> inheritances = new ArrayList<>();
  public final List<String> overrides = new ArrayList<>();
  public final List<String> usages = new ArrayList<>();
  public final List<String> typeUses = new ArrayList<>();
  public final List<String> typeArguments = new ArrayList<>();
  public final List<String> imports = new ArrayList<>();
  public final List<String> annotationUses = new ArrayList<>();

  public final List<String> errors = new ArrayList<>();
  public final Set<String> files = new HashSet<>();

  private TestStorage(IntermediateStorage proto) {
    this.proto = proto;
  }

  public static TestStorage create(IntermediateStorage proto) {
    TestStorage storage = new TestStorage(proto);
    storage.build();
    return storage;
  }

  // ---- build ----------------------------------------------------------

  private void build() {
    for(StorageFile file : proto.getFilesList()) {
      files.add(file.getFilePath());
    }

    Map<Long, Integer> accessByNodeId = new HashMap<>();
    for(StorageComponentAccess access : proto.getComponentAccessesList()) {
      accessByNodeId.put(access.getNodeId(), access.getType());
    }

    Map<Long, List<Long>> elementIdsByLocationId = new HashMap<>();
    for(StorageOccurrence occurrence : proto.getOccurrencesList()) {
      elementIdsByLocationId.computeIfAbsent(occurrence.getSourceLocationId(), k -> new ArrayList<>())
          .add(occurrence.getElementId());
    }

    Map<Long, List<StorageSourceLocation>> tokenLocations = new HashMap<>();
    Map<Long, List<StorageSourceLocation>> scopeLocations = new HashMap<>();
    Map<Long, List<StorageSourceLocation>> signatureLocations = new HashMap<>();
    Map<Long, List<StorageSourceLocation>> localSymbolLocations = new HashMap<>();
    Map<Long, List<StorageSourceLocation>> qualifierLocations = new HashMap<>();
    Map<Long, List<StorageSourceLocation>> errorLocations = new HashMap<>();

    for(StorageSourceLocation location : proto.getSourceLocationsList()) {
      List<Long> elementIds = elementIdsByLocationId.get(location.getId());
      if(elementIds == null || elementIds.isEmpty()) {
        continue;
      }
      Map<Long, List<StorageSourceLocation>> target = switch((int) location.getType()) {
        case Kinds.LOCATION_TOKEN -> tokenLocations;
        case Kinds.LOCATION_SCOPE -> scopeLocations;
        case Kinds.LOCATION_QUALIFIER -> qualifierLocations;
        case Kinds.LOCATION_LOCAL_SYMBOL -> localSymbolLocations;
        case Kinds.LOCATION_SIGNATURE -> signatureLocations;
        case Kinds.LOCATION_ERROR -> errorLocations;
        default -> null;
      };
      if(target == null) {
        continue;
      }
      for(long elementId : elementIds) {
        if(elementId != 0) {
          target.computeIfAbsent(elementId, k -> new ArrayList<>()).add(location);
        }
      }
    }

    Map<Long, StorageNode> nodesById = new HashMap<>();
    Set<Long> fileNodeIds = new HashSet<>();
    for(StorageNode node : proto.getNodesList()) {
      nodesById.put(node.getId(), node);
      if(node.getType() == Kinds.NODE_FILE) {
        fileNodeIds.add(node.getId());
      }

      String nameStr = renderName(node.getSerializedName());

      for(StorageSourceLocation loc : qualifierLocations.getOrDefault(node.getId(), List.of())) {
        qualifiers.add(nameStr + addLocationStr("", loc));
      }

      List<String> bin = binForNodeType(node.getType());
      if(bin == null) {
        continue;
      }

      Integer access = accessByNodeId.get(node.getId());
      if(access != null && access != Kinds.ACCESS_TEMPLATE_PARAMETER && access != Kinds.ACCESS_TYPE_PARAMETER) {
        nameStr = accessKindToString(access) + " " + nameStr;
      }

      List<StorageSourceLocation> tokenLocs = tokenLocations.get(node.getId());
      boolean added = false;
      if(tokenLocs != null) {
        for(StorageSourceLocation tokenLoc : tokenLocs) {
          added = false;
          String locationStr = addLocationStr("", tokenLoc);

          StorageSourceLocation signatureLoc = firstOrNull(signatureLocations.get(node.getId()));
          if(signatureLoc != null && containsLocation(signatureLoc, tokenLoc)) {
            locationStr = addLocationStr(locationStr, signatureLoc);
          }

          for(StorageSourceLocation scopeLoc : scopeLocations.getOrDefault(node.getId(), List.of())) {
            if(containsLocation(scopeLoc, tokenLoc)) {
              bin.add(nameStr + addLocationStr(locationStr, scopeLoc));
              added = true;
            }
          }

          if(!added) {
            bin.add(nameStr + locationStr);
            added = true;
          }
        }
      }

      if(!added) {
        bin.add(nameStr);
      }
    }

    for(StorageEdge edge : proto.getEdgesList()) {
      StorageNode target = nodesById.get(edge.getTargetNodeId());
      StorageNode source = nodesById.get(edge.getSourceNodeId());
      if(target == null || source == null) {
        continue;
      }

      List<String> bin = binForEdgeType(edge.getType());
      if(bin == null) {
        continue;
      }

      String sourceName = renderName(source.getSerializedName());
      if(fileNodeIds.contains(edge.getSourceNodeId())) {
        sourceName = fileName(sourceName);
      }
      String targetName = renderName(target.getSerializedName());
      if(fileNodeIds.contains(edge.getTargetNodeId())) {
        targetName = fileName(targetName);
      }
      String nameStr = sourceName + " -> " + targetName;

      List<StorageSourceLocation> tokenLocs = tokenLocations.get(edge.getId());
      boolean added = false;
      if(tokenLocs != null) {
        for(StorageSourceLocation tokenLoc : tokenLocs) {
          bin.add(nameStr + addLocationStr("", tokenLoc));
          added = true;
        }
      }

      if(!added) {
        bin.add(nameStr);
      }
    }

    for(StorageLocalSymbol localSymbol : proto.getLocalSymbolsList()) {
      List<StorageSourceLocation> locs = localSymbolLocations.get(localSymbol.getId());
      boolean added = false;
      if(locs != null) {
        for(StorageSourceLocation loc : locs) {
          localSymbols.add(localSymbol.getName() + addLocationStr("", loc));
          added = true;
        }
      }
      if(!added) {
        localSymbols.add(localSymbol.getName());
      }
    }

    for(StorageError error : proto.getErrorsList()) {
      for(StorageSourceLocation loc : errorLocations.getOrDefault(error.getId(), List.of())) {
        errors.add(error.getMessage() + addLocationStr("", loc));
      }
    }
  }

  // ---- formatting helpers ---------------------------------------------

  /** Renders a {@code Names}-serialized name the way {@code NameHierarchy::getQualifiedNameWithSignature} does. */
  private static String renderName(String serialized) {
    int mIdx = serialized.indexOf("\tm");
    String delimiter = serialized.substring(0, mIdx);
    String[] chunks = serialized.substring(mIdx + 2).split("\tn", -1);
    List<String> names = new ArrayList<>(chunks.length);
    String prefix = "";
    String postfix = "";
    for(int i = 0; i < chunks.length; i++) {
      String chunk = chunks[i];
      int sIdx = chunk.indexOf("\ts");
      int pIdx = chunk.indexOf("\tp", sIdx);
      names.add(chunk.substring(0, sIdx));
      if(i == chunks.length - 1) {
        prefix = chunk.substring(sIdx + 2, pIdx);
        postfix = chunk.substring(pIdx + 2);
      }
    }
    String qualified = String.join(delimiter, names);
    return prefix + (prefix.isEmpty() ? "" : " ") + qualified + postfix;
  }

  private static String addLocationStr(String inner, StorageSourceLocation loc) {
    return " <" + loc.getStartLine() + ":" + loc.getStartCol() + inner + " " + loc.getEndLine() + ":" +
        loc.getEndCol() + ">";
  }

  private static boolean containsLocation(StorageSourceLocation outLocation, StorageSourceLocation inLocation) {
    if(outLocation.getStartLine() > inLocation.getStartLine()) {
      return false;
    }
    if(outLocation.getStartLine() == inLocation.getStartLine() && outLocation.getStartCol() > inLocation.getStartCol()) {
      return false;
    }
    if(outLocation.getEndLine() < inLocation.getEndLine()) {
      return false;
    }
    if(outLocation.getEndLine() == inLocation.getEndLine() && outLocation.getEndCol() < inLocation.getEndCol()) {
      return false;
    }
    return true;
  }

  /** Port of {@code FilePath::fileName()} for the one thing we need it for: the last path segment. */
  private static String fileName(String path) {
    int idx = Math.max(path.lastIndexOf('/'), path.lastIndexOf('\\'));
    return idx < 0 ? path : path.substring(idx + 1);
  }

  /** Port of {@code accessKindToString} (AccessKind.cpp). */
  private static String accessKindToString(int access) {
    return switch(access) {
      case Kinds.ACCESS_PUBLIC -> "public";
      case Kinds.ACCESS_PROTECTED -> "protected";
      case Kinds.ACCESS_PRIVATE -> "private";
      case Kinds.ACCESS_DEFAULT -> "default";
      default -> "";
    };
  }

  private List<String> binForNodeType(int type) {
    if(type == Kinds.NODE_PACKAGE) {
      return packages;
    }
    if(type == Kinds.NODE_CLASS) {
      return classes;
    }
    if(type == Kinds.NODE_INTERFACE) {
      return interfaces;
    }
    if(type == Kinds.NODE_ANNOTATION) {
      return annotations;
    }
    if(type == Kinds.NODE_ENUM) {
      return enums;
    }
    if(type == Kinds.NODE_ENUM_CONSTANT) {
      return enumConstants;
    }
    if(type == Kinds.NODE_METHOD) {
      return methods;
    }
    if(type == Kinds.NODE_FIELD) {
      return fields;
    }
    if(type == Kinds.NODE_FUNCTION) {
      return functions;
    }
    if(type == Kinds.NODE_TYPE_PARAMETER) {
      return typeParameters;
    }
    if(type == Kinds.NODE_NAMESPACE) {
      return namespaces;
    }
    return null;
  }

  private List<String> binForEdgeType(int type) {
    if(type == Kinds.EDGE_CALL) {
      return calls;
    }
    if(type == Kinds.EDGE_INHERITANCE) {
      return inheritances;
    }
    if(type == Kinds.EDGE_OVERRIDE) {
      return overrides;
    }
    if(type == Kinds.EDGE_USAGE) {
      return usages;
    }
    if(type == Kinds.EDGE_TYPE_USAGE) {
      return typeUses;
    }
    if(type == Kinds.EDGE_TYPE_ARGUMENT) {
      return typeArguments;
    }
    if(type == Kinds.EDGE_IMPORT) {
      return imports;
    }
    if(type == Kinds.EDGE_ANNOTATION_USAGE) {
      return annotationUses;
    }
    return null;
  }

  private static <T> T firstOrNull(List<T> list) {
    return (list == null || list.isEmpty()) ? null : list.get(0);
  }

  // ---- proto-level helpers (lifted from JavaIndexerTest) ---------------

  public StorageNode node(int type, String nameContains) {
    return proto.getNodesList().stream()
        .filter(n -> n.getType() == type && n.getSerializedName().contains(nameContains))
        .findFirst()
        .orElseThrow(() -> new AssertionError(
            "no node type=" + type + " containing \"" + nameContains + "\" in " + names()));
  }

  public boolean hasNode(int type) {
    return proto.getNodesList().stream().anyMatch(n -> n.getType() == type);
  }

  public boolean hasEdge(long from, long to, int type) {
    return proto.getEdgesList().stream()
        .anyMatch(e -> e.getSourceNodeId() == from && e.getTargetNodeId() == to && e.getType() == type);
  }

  public boolean hasEdgeFrom(long from, int type) {
    return proto.getEdgesList().stream().anyMatch(e -> e.getSourceNodeId() == from && e.getType() == type);
  }

  public List<StorageEdge> edgesOf(int type) {
    return proto.getEdgesList().stream().filter(e -> e.getType() == type).toList();
  }

  /** Location types attached to an element id, via the occurrence table. */
  public Set<Integer> locationTypesOf(long elementId) {
    Set<Long> locIds = proto.getOccurrencesList().stream()
        .filter(o -> o.getElementId() == elementId)
        .map(StorageOccurrence::getSourceLocationId)
        .collect(HashSet::new, Set::add, Set::addAll);
    return proto.getSourceLocationsList().stream()
        .filter(l -> locIds.contains(l.getId()))
        .map(l -> (int) l.getType())
        .collect(HashSet::new, Set::add, Set::addAll);
  }

  public boolean hasSymbol(long id) {
    return proto.getSymbolsList().stream().anyMatch(sym -> sym.getId() == id);
  }

  public int accessOf(long id) {
    return proto.getComponentAccessesList().stream()
        .filter(a -> a.getNodeId() == id)
        .map(a -> (int) a.getType())
        .findFirst()
        .orElse(Kinds.ACCESS_NONE);
  }

  /** Raw serialized names, for diagnostics -- {@code AssertionError} messages, "must appear once" checks. */
  public List<String> names() {
    return proto.getNodesList().stream().map(StorageNode::getSerializedName).toList();
  }

  public IntermediateStorage proto() {
    return proto;
  }
}
